File: gradle-plugin/src/main/java/org/komapper/gradle/KomapperExtension.java
package org.komapper.gradle;

import org.gradle.api.Action;
import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.Project;
import org.komapper.gradle.codegen.Generator;

import java.util.Objects;

public class KomapperExtension {
    public static final String NAME = "komapper";
    private final NamedDomainObjectContainer<Generator> generators;

    public KomapperExtension(Project project) {
        Objects.requireNonNull(project);
        this.generators =
                project.container(
                        Generator.class,
                        name -> project.getObjects().newInstance(Generator.class, name, project));
    }

    public NamedDomainObjectContainer<Generator> getGenerators() {
        return generators;
    }

    @SuppressWarnings("unused")
    public void generators(Action<? super NamedDomainObjectContainer<Generator>> action) {
        action.execute(generators);
    }
}

File: gradle-plugin/src/main/kotlin/org/komapper/gradle/KomapperExtension.kt
package org.komapper.gradle

import org.gradle.api.Action
import org.gradle.api.Project
import org.komapper.gradle.codegen.Generator
import org.komapper.gradle.codegen.GeneratorContainer
import org.komapper.gradle.codegen.GeneratorContainerImpl
import javax.inject.Inject

open class KomapperExtension @Inject internal constructor(private val project: Project) {
    companion object {
        const val NAME = "komapper"
    }

    internal val generators: GeneratorContainer =
        GeneratorContainerImpl(
            project.container(Generator::class.java) { name ->
                project.objects.newInstance(Generator::class.java, name, project)
            }
        )

    fun generators(action: Action<in GeneratorContainer>) {
        action.execute(generators)
    }
}
